### PR TITLE
Refactored & Added Unit Tests

### PR DESCRIPTION
**Pull request by Hopding**
## Description

In this PR I have:

* Relocated & refactored some files.
* Sprinkled in some magic pixie dust ✨  (ES6 syntax) to improve readability.
* Renamed some folders.
* Updated the build/compile process to be run in its entirety every time `make` is run, thereby removing the quirkiness of the `_cache` directory.
* Added some unit tests for the transformers.

## Motivation and Context

Addresses #61. Note that more testing (& potentially refactoring) remains to be done before closing that issue, but I think it will be best to merge them in pieces rather than all at once.

## Checklist:

- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by thejoelw — APPROVED** [review]
Looks good so far. It's a partial refactor, so don't merge into master.

**Comment by thebho** [comment]
If this is working, I think we should merge now.

File: src/compile.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+const resolve_grammar_path = (lang_key, grammar_file) =>
+      path.resolve(__dirname, '..', 'grammars-v4', lang_key, grammar_file);
+
+/*
+Sequentially invokes each task in `tasks` using the previous tasks's
+return value as its argument. `tasks` must be an array of async functions.
+The first task is invoked with a null argument.
+*/
+const waterfall = async (tasks) => {
```
Comment (dane-johnson) [inline src/compile.js:19]: Not sure about this function name.

File: utils/sample-ast.json [hunk 2]
```diff
@@ -0,0 +1,277 @@
+{
```
Comment (dane-johnson) [inline utils/sample-ast.json]: What is this for?

File: utils/utils.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
+const { TerminalNodeImpl } = require('antlr4/tree/Tree.js');
```
Comment (dane-johnson) [inline utils/utils.js]: `utils/utils.js` is confusing. If this is the default entry for this module, should be named `index.js`

**Comment by thejoelw** [comment]
Was there a reason this PR was closed?

**Comment by thebho** [comment]
I honestly don't remember

**Pull request merged**